libkbfs/bare_root_metadata_v3.go

```diff
@@ -1013,21 +1040,71 @@ func (md *BareRootMetadataV3) GetTLFReaderKeyBundleID() TLFReaderKeyBundleID {
 }
 
 // FinalizeRekey implements the MutableBareRootMetadata interface for BareRootMetadataV3.
-func (md *BareRootMetadataV3) FinalizeRekey(config Config, extra ExtraMetadata) error {
+func (md *BareRootMetadataV3) FinalizeRekey(crypto cryptoPure, prevKey, currKey TLFCryptKey,
+	extra ExtraMetadata) error {
 	extraV3, ok := extra.(*ExtraMetadataV3)
 	if !ok {
 		return errors.New("Invalid extra metadata")
 	}
+	if md.LatestKeyGeneration() < KeyGen(FirstValidKeyGen) {
+		return fmt.Errorf("Invalid key generation %d", md.LatestKeyGeneration())
+	}
+	if (prevKey != TLFCryptKey{}) {
+		numKeys := int(md.LatestKeyGeneration() - 1)
+		var oldKeys []TLFCryptKey
+		var err error
+		if numKeys > 1 {
+			oldKeys, err = crypto.DecryptTLFCryptKeys(
+				extraV3.wkb.EncryptedHistoricTLFCryptKeys, prevKey)
+			if err != nil {
+				return err
+			}
+		}
+		oldKeys = append(oldKeys, prevKey)
+		encOldKeys, err := crypto.EncryptTLFCryptKeys(oldKeys, currKey)
+		if err != nil {
+			return err
+		}
+		extraV3.wkb.EncryptedHistoricTLFCryptKeys = encOldKeys
+	}
 	var err error
-	md.WriterMetadata.WKeyBundleID, err =
-		config.Crypto().MakeTLFWriterKeyBundleID(extraV3.wkb)
+	md.WriterMetadata.WKeyBundleID, err = crypto.MakeTLFWriterKeyBundleID(extraV3.wkb)
 	if err != nil {
 		return err
 	}
-	md.RKeyBundleID, err = config.Crypto().MakeTLFReaderKeyBundleID(extraV3.rkb)
+	md.RKeyBundleID, err = crypto.MakeTLFReaderKeyBundleID(extraV3.rkb)
 	return err
 }
 
+// StoresHistoricTLFCryptKeys implements the BareRootMetadata interface for BareRootMetadataV3.
+func (md *BareRootMetadataV3) StoresHistoricTLFCryptKeys() bool {
+	return true
+}
+
+// GetHistoricTLFCryptKey implements the BareRootMetadata interface for BareRootMetadataV3.
+func (md *BareRootMetadataV3) GetHistoricTLFCryptKey(crypto cryptoPure,
+	keyGen KeyGen, currentKey TLFCryptKey, extra ExtraMetadata) (TLFCryptKey, error) {
+	extraV3, ok := extra.(*ExtraMetadataV3)
+	if !ok {
+		return TLFCryptKey{}, errors.New("Invalid extra metadata")
+	}
+	if md.LatestKeyGeneration() < KeyGen(FirstValidKeyGen) ||
```

Comment:
The first clause should be `keyGen < KeyGen(FirstValidKeyGen)` -- we want `FirstValidKeyGen <= keyGen < md.LatestKeyGeneration()` to hold.